File: solutions/2026-07-16/solution.py
# Validate IP Address
# Difficulty: Medium
# URL: https://leetcode.com/problems/validate-ip-address/

def restoreIpAddresses(s: str) -> [str]:
  def is_valid(seg: str) -> bool:
    if len(seg) > 1 and seg.startswith('0'): return False
    return 0 <= int(seg) <= 255

  res = []
  for i in range(1, min(4, len(s))):
    for j in range(i + 1, min(i + 4, len(s))):
      for k in range(j + 1, min(j + 4, len(s))):
        seg1, seg2, seg3, seg4 = s[:i], s[i:j], s[j:k], s[k:]
        if is_valid(seg1) and is_valid(seg2) and is_valid(seg3) and is_valid(seg4):
          res.append('.'.join([seg1, seg2, seg3, seg4]))
  return res

File: solutions/2026-07-16/solution.java
// Validate IP Address
// Difficulty: Medium
// URL: https://leetcode.com/problems/validate-ip-address/

class Solution {
  public List<String> restoreIpAddresses(String s) {
    List<String> res = new ArrayList<>();
    for (int i = 1; i <= Math.min(4, s.length()); i++) {
      for (int j = i + 1; j <= Math.min(i + 4, s.length()); j++) {
        for (int k = j + 1; k <= Math.min(j + 4, s.length()); k++) {
          String seg1 = s.substring(0, i);
          String seg2 = s.substring(i, j);
          String seg3 = s.substring(j, k);
          String seg4 = s.substring(k);
          if (isValid(seg1) && isValid(seg2) && isValid(seg3) && isValid(seg4)) {
            res.add(seg1 + '.' + seg2 + '.' + seg3 + '.' + seg4);
          }
        }
      }
    }
    return res;
  }

  private boolean isValid(String seg) {
    if (seg.length() > 1 && seg.startsWith("0")) return false;
    return 0 <= Integer.parseInt(seg) <= 255;
  }
}